azure-pipelines/templates/steps/vsts-release.yml

```
@@ -0,0 +1,86 @@
+# File: azure-pipelines\templates\steps\nightly-snapshot-releases\vsts-release.yml
+# Description: Template to assemble and publish Snapshot/Latest versions of ${{ parameters.projectName }}
+# Variable: 'ENV_VSTS_MVN_ANDROIDADACCOUNTS_USERNAME' was defined in the Variables tab
+# Variable: 'mvnAccessToken' was defined in the Variables tab
+# Variable: 'folderName' refers to the project folder
+# Variable: 'gradleTasks' list all the gradle tasks to be used.
+# Variable: 'accessTokenName' refers to the name of the access Token in this project.
+# Variable: 'runUnitTest' boolean: refers to whether the pipeline will run unit tests or not.
+# https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate
+# Variable: 'versionType' type of version (Latest, Snapshot)
+
+parameters:
+- name: versionType
+  default: Latest
+  values:
+  - Snapshot
+  - Latest
+- name: folderName
+  default: none
+- name: projectName
+  default: none
+- name: gradleTasks
+  default: none
+- name: accessTokenName
+  default: none
+- name: runUnitTest
+  default: false
+
+jobs:
+- job: Phase_1
+  displayName: Phase 1
+  timeoutInMinutes: 90
+  cancelTimeoutInMinutes: 1
+  pool:
+    name: Hosted Windows 2019 with VS2019
+  steps:
+  - checkout: self
+    clean: true
+    submodules: recursive
+    persistCredentials: True
+  - template: credscan-policheck.yml
+    parameters:
+      policheckCmdLineArgsDir: ${{ parameters.projectName }}
+  - task: CmdLine@1
+    displayName: Set MVN Access Token in Environment
+    inputs:
+      filename: echo
+      arguments: '##vso[task.setvariable variable=${{ parameters.accessTokenName }}]$(mvnAccessToken)'
+  - task: Gradle@2
+    displayName: ${{ parameters.versionType }} version update
+    inputs:
+      tasks: ${{ parameters.projectName }}:version${{ parameters.versionType }}
+      publishJUnitResults: false
+  - task: Gradle@1
+    name: Gradle1
+    displayName: Assemble Release
+    inputs:
+      tasks: ${{ parameters.gradleTasks }}
```

you are passing `clean AADAuthenticator:assembleLatest` as parameters.gradleTasks
so versionType is ignored here, also you are passing the project name that we already have

my recommendation it is to change it this way and remove the parameter gradleTasks
```suggestion
      tasks: clean  ${{ parameters.projectName }}:assemble${{ parameters.versionType }}
```

---

Fair point, however, there is a challenge that comes up. The above suggestion works very well with assemble tasks that have the android gradle plugin e.g. (in AADAuthenticator), when you use this task with the java-library gradle plugin, <project>:assemble task has no variants specifically targeting the flavour.

---

One thing you may want to try is  add  '  ' to parameters.versionType or make it default to empty :)

---

parameters.versionType is a mandatory field. Secondly, I think the specifying the gradleTasks directly is more preferrable as compared to having the tasks be resolved. There are instances one may want to build using different tasks e.g. `<projectName>:build` over `<projectName:>assemble` and also assists on the readability of the caller script.